File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        dp, buy = [0] * len(prices), prices[0]
        
        for i, p in enumerate(prices):
            buy = min(buy, p)
            dp[i] = max(prices[i] - buy, dp[i-1])
        return dp[-1]

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int ans = 0;
    int dp(int idx, int prev, vector<int>& prices) {
        int n = prices.size();
        if (idx >= n) return 0;
        ans = max(ans, prices[idx] - prev);
        dp(idx + 1, min(prev, prices[idx]), prices);
        return ans;
    }
    int maxProfit(vector<int>& prices) {
        return dp(0, prices[0], prices);
    }
};